### ISSUE TITLE
Callback-based canvas source (or layer) API

### ISSUE THREAD

**Issue by anandthakker** [mapbox/mapbox-gl-js#5557]
[CanvasSource](https://www.mapbox.com/mapbox-gl-js/api/#canvassource) allows users to create raster layers that use the contents of an arbitrary HTML Canvas element.  However, like [ImageSource](https://www.mapbox.com/mapbox-gl-js/api/#imagesource), this API is currently oriented towards drawing into the fixed geographical area (determined by the `coordinates` option).   This works well for drawing (possibly animated) graphics in a specific place, but it's not well suited to drawing graphics that "follow" the map, responding in realtime to camera movement.

To that end, I tentatively propose that we provide an API that allows users to provide a `render: (CameraState) => void` callback in addition to an HTML Canvas element.  Our rendering pipeline would call the callback on every frame before reading data from the canvas.  The canvas data would be assumed to be coincident with the map's viewport.

We could expose such an API by:

1. _Extending the `CanvasSource` API:_ Allow users to specify _either_ `coordinates:` _or_ `render: (CameraState) => void`.
2. _Adding a new source type:_ Like 1, but make it a separate source type.
3. _Adding a new layer type:_ Options 1 and 2 treat the canvas as a _source_ that can be targeted by `raster` layers.  Instead, we could expose a `CanvasLayer` that provides the canvas element and `render` callback.

**Comment by asheemmamoowala** [comment]
It doesn't seem right to have a _source_ type responsible for directly drawing to the map. A `CanvasLayer` is more appropriate - it would be similar to the `CustomLayer` in Native.
Another option would be to create a `CustomLayer` type now with a render callback that can be implemented via `canvas` or OpenGL (in the future?)

**Comment by anandthakker** [comment]
> It doesn't seem right to have a source type responsible for directly drawing to the map

I agree, but the existing CanvasSource doesn't draw to the map either: it draws to a canvas that is then used by one or more raster layers.  A CanvasSource variation like the one described here _could_ work the same way (and thus allow the raster layer using it to manipulate what's actually drawn using the various `raster-*` style properties).  I'm not actually sure if that's desirable, though -- just noting that it's an option.

**Comment by asheemmamoowala** [comment]
I just don't see the benefits of the `CanvasSource` + `RasterLayer` variation. All the `raster-*` style property changes can be baked into the `canvas` drawing especially since there isn't any data-driven styling. 
I see the benefit of reusing that combination for the short term. In the long term though it would be preferrable to replace `CanvasSource` from the `style-spec` and move to a `CustomLayer`-like API.

**Comment by anandthakker** [comment]
Yeah, I find that convincing @asheemmamoowala -- locating this API (these APIs) in the realm of layers feels right

**Comment by kronick** [comment]
@anandthakker @asheemmamoowala  Love this proposal. I have been using a workaround for this, setting the `canvas` `coords` to match the viewport bounds on map `move` events, and matching the Mapbox map projection and viewport transformation manually. Here's an example: https://jsfiddle.net/kronick/kzxn5dsb/

![](https://user-images.githubusercontent.com/186834/33402985-67b2583e-d513-11e7-962d-caa010b9c42a.gif)

What properties would you imagine `CameraState` having? I'm using undocumented properties in `map.transform` that would be nice to have officially supported. 

Another way to do this might be to add a `raster-alignment` layout property that could be either `viewport` or `map`-- if it's `viewport`, the raster layer would render to always fill the entire viewport and ignore its `coordinates` while `map` would render pinned to a geographic location as it is now.

You can already simulate the callback behavior by registering a callback for map `move` or `render` events. This is the approach we took in `threebox`: https://github.com/peterqliu/threebox/blob/master/src/Camera/CameraSync.js#L20 . Synchronizing this to only re-render when the map was re-drawing would be 💯 .

**Comment by kronick** [comment]
Here's another example I made a while back that lets you toggle between filling the viewport and pinning a canvas to geographic coords: https://www.mapbox.com/labs/canvas-heatmap/debug.html

Note that the workaround totally breaks when you pitch the map.
